packages/dds/tree/src/core/checkout/branch.ts

```diff
@@ -0,0 +1,155 @@
+}
+
+/**
+ * A branch of changes that can be applied to a SharedTree.
+ */
+export class SharedTreeBranch<TChange> extends EventEmitter<SharedTreeBranchEvents<TChange>> {
+	private head: GraphCommit<TChange>;
+	private readonly forks = new Set<SharedTreeBranch<TChange>>();
+	private disposed = false;
+
+	/**
+	 * @param getBaseBranch - a function which retrieves the head of the base branch
+	 * @param mergeIntoBase - a function which describes how to merge this branch into its base branch
+	 * @param sessionId - the session ID used to author commits made by to this branch
+	 * @param rebaser - a rebaser to rebase this branch's changes when it pulls or merges
+	 */
+	public constructor(
+		private readonly getBaseBranch: () => GraphCommit<TChange>,
+		private readonly mergeIntoBase: (forked: SharedTreeBranch<TChange>) => TChange,
+		private readonly sessionId: string,
+		private readonly rebaser: Rebaser<TChange>,
+	) {
+		super();
+		this.head = getBaseBranch();
+	}
+
+	/**
+	 * @returns the commit at the head of this branch.
+	 */
+	public getHead(): GraphCommit<TChange> {
+		return this.head;
+	}
+
+	/**
+	 * Apply the given change to this branch.
+	 * Emits an `onChange` event.
+	 * @returns the net change to this branch (i.e. `change`)
+	 */
+	public applyChange(change: TChange): TChange {
+		this.assertNotDisposed();
+		this.head = mintCommit(this.head, {
+			revision: mintRevisionTag(),
+			sessionId: this.sessionId,
+			change,
+		});
+		this.emit("onChange", change);
+		return change;
+	}
+
+	/**
+	 * Rebase the changes that have been applied to this branch over all the changes in the base branch that have
+	 * occurred since this branch last pulled (or was forked).
+	 * @returns the net change to this branch
+	 */
+	public pull(): TChange {
+		this.assertNotDisposed();
+		const baseBranch = this.getBaseBranch();
+		if (this.head === baseBranch) {
+			// Not necessary for correctness, but skips needless rebase and event firing below
+			return this.rebaser.changeRebaser.compose([]);
+		}
+
+		const [newBranch, change] = this.rebaser.rebaseBranch(this.head, baseBranch);
+		this.head = newBranch;
+		this.emit("onChange", change);
+		return change;
+	}
+
+	/**
+	 * Spawn a new branch that is based off of the current state of this branch.
+	 * Changes made to the new branch will not be applied to this branch until the new branch is merged back in.
+	 */
+	public fork(): SharedTreeBranch<TChange> {
+		this.assertNotDisposed();
+		const fork = new SharedTreeBranch(
+			() => this.head,
+			(forked) => {
+				// In this function, `this` is the base and `forked` is the fork being merged in
+				const forkedChange = forked.pull();
+				const commits: GraphCommit<TChange>[] = [];
+				const baseBranch = forked.getBaseBranch();
+				const foundBaseBranch = findAncestor(
+					[forked.head, commits],
+					(c) => c === baseBranch,
+				);
+				assert(
+					foundBaseBranch !== undefined,
+					"Expected merging checkout branches to be related",
```

Comment:
Shouldn't the ancestor always be `this.head` since we just did a pull on the fork?

Reply:
Yes, nice.